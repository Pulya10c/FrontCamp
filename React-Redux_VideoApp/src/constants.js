export const SM_SCREEN_SIZE = 576;
export const MD_SCREEN_SIZE = 768;

export const PATHS = {
  ANY: '*',
  HOME: '/',
  VIDEO_LIST_PAGE: '/search',
  CURRENT_VIDEO_PAGE: '/film/:id'
};

export const API = {
  SEARCH: 'search',
  SEARCH_BY: 'searchBy',
  SORT_ORDER: 'sortOrder',
  SORT_BY: 'sortBy',
  OFFSET: 'offset',
  LIMIT: 'limit'
};

export const TITLE = { text: 'Title', code: 'title' };
export const GENRES = { text: 'Genres', code: 'genres' };

export const RELEASE_DATE = { text: 'Release date', code: 'release_date' };
export const RATING = { text: 'Rating', code: 'vote_average' };

export const ASCENDING = 'asc';
export const DESCENDING = 'desc';

export const SORT_BY_VALUES = [RELEASE_DATE, RATING];
export const SEARCH_BY_VALUES = [TITLE, GENRES];
export const SORT_ORDER_VALUES = [ASCENDING, DESCENDING];
