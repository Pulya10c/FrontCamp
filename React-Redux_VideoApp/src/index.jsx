import React from "react";
import ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { ConnectedRouter } from 'connected-react-router'

import ErrorBoundary from "components/ErrorBoundary";
import App from "./App";
import { store, history } from "./configureStore";
import "./style.scss";

ReactDOM.render(
  <ErrorBoundary>
    <Provider store={store}>
      <App />
    </Provider>
  </ErrorBoundary>,
  document.getElementById("root")
);
