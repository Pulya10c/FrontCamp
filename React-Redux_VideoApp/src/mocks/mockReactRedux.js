const createMockMapStateToProps = mapStateToProps => {
  if (!mapStateToProps) {
    return mapStateToProps;
  }

  return (state, ownProps) => {
    return {
      props: mapStateToProps(state, ownProps)
    };
  };
};

const getDispatchFunctionFormShorthandObject = actionsObj => dispatch => {
  return Object.keys(actionsObj).reduce((props, actionKey) => {
    return {
      ...props,
      [actionKey]: function() {
        return dispatch(actionsObj[actionKey].apply(this, arguments));
      }
    };
  }, {});
};

const createMockMapDispatchToProps = mapDispatchToProps => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps;
  }

  const mappingFunction =
    typeof mapDispatchToProps === 'function'
      ? mapDispatchToProps
      : getDispatchFunctionFormShorthandObject(mapDispatchToProps);

  return (dispatch, ownProps) => {
    dispatch = dispatch || jest.fn(action => action);

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    };
  };
};

const createMockMergeProps = mergeProps => {
  if (!mergeProps) {
    return mergeProps;
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: jest.fn(action => action) };

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    };
  };
};

export default {
  connect: (
    mapStateToProps,
    mapDispatchToProps,
    mergeProps
  ) => ConnectedComponent => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    ConnectedComponent
  })
};
