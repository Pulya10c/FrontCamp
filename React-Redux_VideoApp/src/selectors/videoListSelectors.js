import { createSelector } from 'reselect'
import get from 'lodash/get';

export const videoListSelector = state => get(state, 'videoList');

export const filterConfigSelector = createSelector(
  [videoListSelector],
  videoList => get(videoList, 'filterConfig')
);

export const searchSelector = createSelector(
  [filterConfigSelector],
  filterConfig => get(filterConfig, 'search')
);

export const searchBySelector = createSelector(
  [filterConfigSelector],
  filterConfig => get(filterConfig, 'searchBy')
);
