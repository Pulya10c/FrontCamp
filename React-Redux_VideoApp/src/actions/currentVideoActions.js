import QueryStringGenerator from 'services/QueryStringGenerator';
import RequestService from 'services/RequestService';
import { updateFilterConfigBy, fetchMovies } from 'actions/videoListActions';
import { API, GENRES, TITLE, PATHS } from 'src/constants';
import { currentVideoGenresSelector } from 'selectors/currentVideoSelectors';
import * as uiActions from 'actions/uiStateActions';
import { goTo } from 'actions/routerActions';

const MODULE_NAME = 'CURRENT_VIDEO';

export const SAVE_FILM = `${MODULE_NAME}/SAVE_FILM`;
export const SET_SELECTED_FILM_ID = `${MODULE_NAME}/SET_SELECTED_FILM_ID`;

// export for test
export const _saveFilm = film => {
  return {
    type: SAVE_FILM,
    payload: film
  };
};

export const fetchCurrentVideoAndSimilarFilms = id => async (
  dispatch,
  getState
) => {
  try {
    const queryString = QueryStringGenerator.getUrlForRequest(id);
    dispatch(uiActions.startFetching());
    const film = await RequestService.getResponse(queryString);
    dispatch(uiActions.finishFetching());
    dispatch(_saveFilm(film));
    const state = getState();
    const genres = currentVideoGenresSelector(state);
    dispatch(updateFilterConfigBy[API.SEARCH](genres.join(' ')));
    dispatch(updateFilterConfigBy[API.SEARCH_BY](GENRES));
    dispatch(fetchMovies());
  } catch (error) {
    console.error(error);
    dispatch(uiActions.setError(error));
  }
};

export const resetSearchAngRedirect = () => dispatch => {
  dispatch(updateFilterConfigBy[API.SEARCH](''));
  dispatch(updateFilterConfigBy[API.SEARCH_BY](TITLE.code));
  goTo(PATHS.VIDEO_LIST_PAGE);
};
