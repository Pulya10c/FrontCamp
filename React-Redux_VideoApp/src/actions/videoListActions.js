import has from 'lodash/has';

import { filterConfigSelector } from 'selectors/videoListSelectors';
import QueryStringGenerator from 'services/QueryStringGenerator';
import RequestService from 'services/RequestService';
import * as uiActions from 'actions/uiStateActions';
import { goTo } from 'actions/routerActions';
import { routerSearchSelector } from 'selectors/routerSelectors';
import { PATHS, API, GENRES } from 'src/constants';

const MODULE_NAME = 'VIDEO_LIST';

export const SET_SEARCH_VALUE = `${MODULE_NAME}/SET_SEARCH_VALUE`;
export const SET_SEARCH_BY_VALUE = `${MODULE_NAME}/SET_SEARCH_BY_VALUE`;
export const SET_SORT_ORDER_VALUE = `${MODULE_NAME}/SET_SORT_ORDER_VALUE`;
export const SET_SORT_BY_VALUE = `${MODULE_NAME}/SET_SORT_BY_VALUE`;
export const SET_OFFSET_VALUE = `${MODULE_NAME}/SET_OFFSET_VALUE`;
export const SET_LIMIT_VALUE = `${MODULE_NAME}/SET_LIMIT_VALUE`;

export const SAVE_MOVIES = `${MODULE_NAME}/SAVE_MOVIES`;
export const UPDATE_FILTER_FROM_SEARCH_PARAMS = `${MODULE_NAME}/UPDATE_FILTER_FROM_SEARCH_PARAMS`;

export const updateFilterConfigBy = {
  [API.SEARCH_BY]: value => ({
    type: SET_SEARCH_BY_VALUE,
    payload: value
  }),
  [API.SEARCH]: value => ({
    type: SET_SEARCH_VALUE,
    payload: value
  }),
  [API.SORT_ORDER]: value => ({
    type: SET_SORT_ORDER_VALUE,
    payload: value
  }),
  [API.SORT_BY]: value => ({
    type: SET_SORT_BY_VALUE,
    payload: value
  }),
  [API.OFFSET]: value => ({
    type: SET_OFFSET_VALUE,
    payload: value
  }),
  [API.LIMIT]: value => ({
    type: SET_LIMIT_VALUE,
    payload: value
  })
};

// export for tests
export const _saveMovies = movies => {
  return {
    type: SAVE_MOVIES,
    payload: movies
  };
};

const _updateFilterFromSearchParams = () => ({
  type: UPDATE_FILTER_FROM_SEARCH_PARAMS
});

// export for tests
export const _updateFilterConfigInOrderToSearchParams = (
  searchStr,
  filterConfig
) => dispatch => {
  const searchParams = new URLSearchParams(searchStr);
  for (const [key, value] of searchParams) {
    if (has(filterConfig, key) && filterConfig[key] != value) {
      dispatch(_updateFilterFromSearchParams());
      dispatch(updateFilterConfigBy[key](value));
    }
  }
};

export const fetchMovies = () => async (dispatch, getState) => {
  try {
    const state = getState();
    const searchStr = routerSearchSelector(state);
    const filterConfig = filterConfigSelector(state);
    dispatch(_updateFilterConfigInOrderToSearchParams(searchStr, filterConfig));
    const queryString = QueryStringGenerator.getUrlForRequest(searchStr);
    dispatch(uiActions.startFetching());
    const movies = await RequestService.getResponse(queryString);
    dispatch(uiActions.finishFetching());
    dispatch(_saveMovies(movies));
  } catch (error) {
    console.error(error);
    dispatch(uiActions.setError(error));
  }
};

export const createSearchAndRedirect = () => (dispatch, getState) => {
  const state = getState();
  const filterConfig = filterConfigSelector(state);
  const searchStr = QueryStringGenerator.getSearchStr(filterConfig);
  goTo(PATHS.VIDEO_LIST_PAGE, searchStr);
};

export const sortByValueAndFetchMovies = value => dispatch => {
  dispatch(updateFilterConfigBy[API.SORT_BY](value));
  dispatch(createSearchAndRedirect());
};
