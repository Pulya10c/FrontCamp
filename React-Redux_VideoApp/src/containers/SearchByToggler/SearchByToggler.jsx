import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import { SEARCH_BY_VALUES } from "src/constants";
import Toggler from "components/Toggler";
import { searchBySelector } from "selectors/videoListSelectors";
import { updateFilterConfigBy } from "actions/videoListActions";
import { API } from "src/constants";

const mapStateToProps = state => ({
  searchByValue: searchBySelector(state)
});

const mapDispatchToProps = dispatch => ({
  setSearchByValue: value =>
    dispatch(updateFilterConfigBy[API.SEARCH_BY](value))
});

const SearchByToggler = ({ searchByValue, setSearchByValue }) => {
  return (
    <Toggler
      togglerTitle="SEARCH BY"
      selectedValue={searchByValue}
      togglerOptions={SEARCH_BY_VALUES}
      setToggleValue={setSearchByValue}
    />
  );
};

SearchByToggler.propTypes = {
  searchByValue: PropTypes.string,
  setSearchByValue: PropTypes.func
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchByToggler);
