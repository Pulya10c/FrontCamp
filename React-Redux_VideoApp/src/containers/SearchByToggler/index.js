import SearchByToggler from './SearchByToggler';

export default SearchByToggler;
