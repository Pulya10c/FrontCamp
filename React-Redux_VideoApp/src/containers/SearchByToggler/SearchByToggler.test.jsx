import React from "react";
import { shallow } from "enzyme";
import mockReactRedux from "mocks/mockReactRedux";
import { searchBySelector } from "selectors/videoListSelectors";
import { updateFilterConfigBy } from "actions/videoListActions";
import { API } from "src/constants";

import SearchByToggler from "./SearchByToggler";

jest.mock("react-redux", () => mockReactRedux);

const mockSearchByValue = "mockSearchByValue";

jest.mock("selectors/videoListSelectors", () => ({
  searchBySelector: jest.fn(() => mockSearchByValue)
}));

const mockUpdateFilterConfigBy = "mockUpdateFilterConfigBy";

jest.mock("actions/videoListActions", () => ({
  updateFilterConfigBy: {
    'searchBy': jest.fn(() => mockUpdateFilterConfigBy)
  }
}));

describe("Component SearchByToggler", () => {
  const {
    mapStateToProps,
    mapDispatchToProps,
    ConnectedComponent
  } = SearchByToggler;

  it("should render correctly", () => {
    const { props } = mapStateToProps();
    expect(shallow(<ConnectedComponent {...props} />)).toMatchSnapshot();
  });

  it("should pass correct props to component from mapStateToProps", () => {
    const { props } = mapStateToProps();
    expect(props).toEqual({
      searchByValue: mockSearchByValue
    });
  });

  it("should pass correct props to component from mapDispatchToProps", () => {
    const { props, dispatch } = mapDispatchToProps();
    const testValue = 'test';
    props.setSearchByValue(testValue);
    expect(updateFilterConfigBy[API.SEARCH_BY]).toHaveBeenCalledWith(testValue);
    expect(dispatch).toHaveBeenCalledWith(mockUpdateFilterConfigBy);
  });
});
