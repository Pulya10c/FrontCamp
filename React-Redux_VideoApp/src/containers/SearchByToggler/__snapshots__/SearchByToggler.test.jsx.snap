// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SearchByToggler should render correctly 1`] = `
<Toggler
  selectedValue="mockSearchByValue"
  togglerOptions={
    Array [
      Object {
        "code": "title",
        "text": "Title",
      },
      Object {
        "code": "genres",
        "text": "Genres",
      },
    ]
  }
  togglerTitle="SEARCH BY"
/>
`;
